src/job-cache/job-cache.cpp

```diff
@@ -759,33 +901,51 @@ AddJobRequest::AddJobRequest(const JAST &job_result_json) {
   //       need this loop. Since this job was just run, wake
   //       will eventually hash all these files so the fact
   //       that we have to re-hash them here is a shame.
-  // TODO: It is incorrect to do this all async without signal
-  //       handling being done correctly.
   // TODO: This code does not handle directories or symlinks correctly
-  std::vector<std::future<OutputFile>> future_outputs;
-
+  // TODO: Use aio_read and do these hashes online and interleaveed
+  //       so that the IO can be in parallel despite the hasing being
+  //       serial.s. It would also be nice to figure out how to do
+  //       the hashing in parallel if we can't avoid it completely.
   // Read the output files which requires kicking off a hash
   for (const auto &output_file : job_result_json.get("output_files").children) {
-    struct stat buf;
+    struct stat buf = {};
     const std::string &src = output_file.second.get("src").value;
-    if (stat(src.c_str(), &buf) < 0) {
-      log_fatal("stat(%s): %s", src.c_str(), strerror(errno));
+    if (lstat(src.c_str(), &buf) < 0) {
+      log_fatal("lstat(%s): %s", src.c_str(), strerror(errno));
     }
-    if (buf.st_mode & S_IFDIR) continue;
-    future_outputs.emplace_back(std::async([&]() {
-      OutputFile output;
-      output.source = output_file.second.get("src").value;
-      output.path = output_file.second.get("path").value;
-      // TODO: This does not work on symlinks or directories right now.
-      auto fd = UniqueFd::open(output.source.c_str(), O_RDONLY);
-      output.hash = do_hash_file(output.source.c_str(), fd.get());
-      return output;
-    }));
-  }
-
-  // Now wait on all the futures
-  for (auto &fut : future_outputs) {
-    outputs.emplace_back(fut.get());
+
+    // Handle output directory
+    if (S_ISDIR(buf.st_mode)) {
+      OutputDirectory dir;
+      dir.mode = buf.st_mode;
+      dir.path = output_file.second.get("path").value;
+      output_dirs.emplace_back(std::move(dir));
+      continue;
+    }
+
+    // Handle symlink
+    if (S_ISLNK(buf.st_mode)) {
+      OutputSymlink sym;
+      static thread_local char link[4097];
+      int size = readlink(src.c_str(), link, sizeof(link));
+      if (size == -1) {
+        log_fatal("readlink(%s): %s", src.c_str(), strerror(errno));
+      }
+      sym.path = output_file.second.get("path").value;
+      sym.value = std::string(link, link + size);
+      output_symlinks.emplace_back(std::move(sym));
+      continue;
+    }
+
+    // Handle regular files but ignore everything else.
+    if (!S_ISREG(buf.st_mode)) continue;
```

Comment:
Do we expect a bunch of things that aren't dirs, links, or regular files? 

if not, might be good to add a debug warning before the continue

Reply:
No I don't think so. The only other options are things like devices, character devices (I think that's a terminal?), block devices, and sockets...stuff that you can't cache even if you wanted to and stuff that fuse isn't going to tell us about.

Reply:
I can and probably should extract this out into a function.

Reply:
I'll not block on it, but I'll strongly urge you to consider it :)

Reply:
Yeah I already did it and its upload now :) I just had stuff to do this morning so it took me a while to finish.